Add support for building CoreCLR for MacCatalyst/iOS simulator

filipnavara opened the pull request:
Re-open and rebase of #98127

Build instructions:
- Build the runtime pack and tools: `./build.sh clr+clr.runtime+libs+packs -os [iossimulator/maccatalyst] -arch [x64/arm64] -cross -c Release`
- Run the sample app: `./dotnet.sh publish src/mono/sample/iOS/Program.csproj -c Release /p:TargetOS=maccatalyst /p:TargetArchitecture=arm64 /p:DeployAndRun=true /p:UseMonoRuntime=false /p:RunAOTCompilation=false /p:MonoForceInterpreter=false`

Related work:
- Counter-part Xamarin changes needed to build full workload and run MAUI apps: https://github.com/xamarin/xamarin-macios/pull/21627 / https://github.com/xamarin/xamarin-macios/compare/net8.0...filipnavara:xamarin-macios:net8.0-coreclr-ioslike?expand=1
- SDK would need to be updated with RIDs in `KnownRuntimeFramework` to use this
- tvOS support is dropped in this version of the PR. The OS doesn't expose Mach exceptions and it would need larger modification in the runtime to add the support.
- We don't use the new JIT API (https://github.com/dotnet/runtime/issues/108423). That's orthogonal problem but would be a necessity to support MacCatalyst/arm64 in production.

Notably, this doesn't include CI scripts to build this or the runtime packs. I am open to suggestions on how to better split this into more digestible/reviewable chunks.

filipnavara committed:
Fix rpath on MacCatalyst

AaronRobinsonMSFT commented on src/coreclr/hosts/inc/coreclrhost.h:
```diff
@@ -19,9 +19,15 @@
 // library and the pointer for explicit dynamic loading (dlopen, LoadLibrary)
+#ifdef __cplusplus
 #define CORECLR_HOSTING_API(function, ...) \
     extern "C" int CORECLR_CALLING_CONVENTION function(__VA_ARGS__); \
```
```suggestion
    EXTERN_C int CORECLR_CALLING_CONVENTION function(__VA_ARGS__); \
```

I think you can just use the `EXTERN_C` macro here.

filipnavara replied:
I'd need to define it in the specific file since it's standalone. That has a risk of collision with other code.

AaronRobinsonMSFT commented on src/coreclr/pal/src/CMakeLists.txt:
```diff
@@ -70,13 +70,18 @@ if(CLR_CMAKE_TARGET_OSX)
+
+if(CLR_CMAKE_TARGET_APPLE)
+  add_definitions(-DTARGET_APPLE)
+  add_definitions(-D_XOPEN_SOURCE)
```
What does this define give us?

filipnavara replied:
I'll have to try again to figure why it was needed here specifically. It changes behavior of some Apple system headers in relation to the POSIX compatibility (https://stackoverflow.com/questions/5378778/what-does-d-xopen-source-do-mean).

filipnavara replied:
Note that we already define this for both NativeAOT and MonoVM:
- https://github.com/dotnet/runtime/blob/2d9f80b0a5f8b4f2c8d4a6fa02f8f8dfcbc26ff5/src/coreclr/nativeaot/CMakeLists.txt#L23
- https://github.com/dotnet/runtime/blob/2d9f80b0a5f8b4f2c8d4a6fa02f8f8dfcbc26ff5/src/mono/CMakeLists.txt#L210

This just aligns CoreCLR on the same plan.

filipnavara replied:
For posterity, the build fails with the following error if `_XOPEN_SOURCE` is not defined:
```
  /Applications/Xcode-16.0.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/usr/include/ucontext.h:51:2: error: The deprecated ucontext routines require _XOPEN_SOURCE to be defined
     51 | #error The deprecated ucontext routines require _XOPEN_SOURCE to be defined
```

AaronRobinsonMSFT replied:
> This just aligns CoreCLR on the same plan.

Perfect. Thanks.

AaronRobinsonMSFT commented on src/coreclr/pal/src/include/pal/dbgmsg.h:
```diff
@@ -494,7 +494,7 @@ if new_level is -1, the nesting level will not be modified
 int DBG_change_entrylevel(int new_level);
 
-#ifdef __APPLE__
+#ifdef TARGET_OSX
```
Is `TARGET_OSX` a subset of `TARGET_APPLE`?

filipnavara replied:
Yes, it is.

AaronRobinsonMSFT commented on src/coreclr/pal/src/map/virtual.cpp:
```diff
@@ -1234,7 +1239,30 @@ VirtualProtect(
+
+#define pthread_jit_write_protect_np jit_write_protect_np
+
+#endif
```
```suggestion
#endif // defined(TARGET_APPLE) && !defined(TARGET_OSX)
```

AaronRobinsonMSFT commented on src/native/corehost/hostpolicy/hostpolicy_context.cpp:
```diff
@@ -70,10 +70,12 @@ namespace
         {
             return CryptoResolveDllImport(entry_point_name);
         }
+#endif
```
```suggestion
#endif // !defined(TARGET_APPLE) || defined(TARGET_OSX)
```

AaronRobinsonMSFT commented on src/native/libs/System.Native/entrypoints.c:
```diff
@@ -46,6 +47,7 @@ static const Entry s_sysNative[] =
     DllImportEntry(SystemNative_ReadStdin)
     DllImportEntry(SystemNative_GetSignalForBreak)
     DllImportEntry(SystemNative_SetSignalForBreak)
+#endif
```
```suggestion
#endif // !defined(TARGET_APPLE) || defined(TARGET_OSX)
```

AaronRobinsonMSFT replied:
I'd like to understand the relationship between `TARGET_APPLE` and `TARGET_OSX`. It would seem to me that OSX should be a subset of APPLE. If so, can't we just say `TARGET_OSX`?

filipnavara replied:
`!defined(TARGET_APPLE) || defined(TARGET_OSX)` is macOS **and Linux/Unix** but not Apple mobile platforms.

matouskozak replied:
We could add `TARGET_APPLE_MOBILE` (we have that one for MonoVM as well) if that would make things more readable.

AaronRobinsonMSFT commented on src/native/libs/System.Security.Cryptography.Native.Apple/entrypoints.c:
```diff
@@ -19,10 +18,16 @@
+#else
+#include "pal_keychain_ios.h"
+#include "pal_x509_ios.h"
+#endif
```
```suggestion
#endif // defined(TARGET_OSX)
```

AaronRobinsonMSFT commented on src/native/libs/System.Security.Cryptography.Native.Apple/entrypoints.c:
```diff
@@ -137,6 +127,24 @@ static const Entry s_cryptoAppleNative[] =
+    DllImportEntry(AppleCryptoNative_X509ExportData)
+    DllImportEntry(AppleCryptoNative_X509CopyWithPrivateKey)
+    DllImportEntry(AppleCryptoNative_X509MoveToKeychain)
+#endif
```
```suggestion
#endif // defined(TARGET_OSX)
```

filipnavara committed:
Apply suggestions from code review

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

janvorli commented on src/native/libs/System.Native/entrypoints.c:
```diff
@@ -34,6 +34,7 @@
 static const Entry s_sysNative[] =
 {
     DllImportEntry(SystemNative_FStat)
+#if !defined(TARGET_APPLE) || defined(TARGET_OSX)
```
It seems that we tend to keep this list the same for all platforms and rather handle the missing functionality in the actual implementation of the methods. E.g. for WASI, many of the methods return EINVAL or NULL.

filipnavara replied:
This is preexisting implementation where it was not done that way. We just never compiled the entrypoints.c on mobile platforms.

Since this is used for corehost which is not particularly useful on mobile platforms maybe I can just drop this part of change and try to exclude corehost build.

filipnavara replied:
Done. There are some additional changes I can revert once https://github.com/dotnet/runtime/pull/110092 and https://github.com/dotnet/runtime/pull/110023 gets in.

filipnavara replied:
Rebased after https://github.com/dotnet/runtime/pull/110092 merge now.

filipnavara commented on src/coreclr/pal/src/map/virtual.cpp:
```diff
@@ -39,12 +39,17 @@ SET_DEFAULT_DEBUG_CHANNEL(VIRTUAL); // some headers have code with asserts, so d
 #include <mach/mach_init.h>
 #endif // HAVE_VM_ALLOCATE
 
+#if defined(TARGET_APPLE) && !defined(TARGET_OSX)
```
You are right. I'll check how the history on why I added it. Hopefully I can just remove it.

filipnavara commented on src/coreclr/pal/src/misc/dbgmsg.cpp:
```diff
@@ -733,7 +733,7 @@ bool DBG_ShouldCheckStackAlignment()
 #endif // _DEBUG && __APPLE__
 
-#ifdef __APPLE__
+#if defined(TARGET_OSX)
```
The code uses the `CFUserNotificationDisplayAlert` API which is not officially available on Mac Catalyst (https://developer.apple.com/documentation/corefoundation/1534496-cfusernotificationdisplayalert?language=objc).

filipnavara commented on src/native/corehost/apphost/static/CMakeLists.txt:
```diff
@@ -176,6 +175,10 @@ else()
     )
 
+    if(NOT CLR_CMAKE_TARGET_MACCATALYST AND NOT CLR_CMAKE_TARGET_IOS AND NOT CLR_CMAKE_TARGET_TVOS AND NOT CLR_CMAKE_HOST_ANDROID)
+        LIST(APPEND NATIVE_LIBS System.Security.Cryptography.Native.OpenSsl-Static)
```
Correct. There's no OpenSSL on iOS and it cannot be globally installed. We never supported local OpenSSL deployments with the app on mobile platforms either.

janvorli replied:
Ok, thank you for the details. But I have one more question, why do we exclude it on Android too?

filipnavara replied:
There's no OpenSSL on Android either. There's a different shim that binds to platform crypto API through JNI.

janvorli replied:
Ah, I didn't know that.

matouskozak commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -188,6 +188,12 @@ add_library(System.Globalization.Native-Static
     entrypoints.c
 )
 
+if (DEFINED CMAKE_ICU_DIR)
```
Looks like we don't need to link the ICU for apple mobile targets, see: https://github.com/dotnet/runtime/blob/c7ada751c4c02996661696ff1f1e1f1328819800/src/native/libs/System.Globalization.Native/CMakeLists.txt#L85-L89

filipnavara committed:
Use CORECLR_HOSTING_API_LINKAGE for extern 'C'

filipnavara committed:
Remove extra #include

filipnavara committed:
Drop ICU linkage (not needed with Hybrid Globalization)

filipnavara committed:
Move jit_write_protect_np initialization

rolfbjarne commented on src/coreclr/dlls/mscoree/coreclr/CMakeLists.txt:
```diff
@@ -228,6 +228,9 @@ endif(CLR_CMAKE_TARGET_WIN32)
 # add the install targets
 install_clr(TARGETS coreclr DESTINATIONS . sharedFramework COMPONENT runtime)
+if(CLR_CMAKE_HOST_MACCATALYST OR CLR_CMAKE_HOST_IOS)
+    install_clr(TARGETS coreclr_static DESTINATIONS . sharedFramework COMPONENT runtime)
```
Linking with a static library typically results in smaller apps, which is why we've always linked Mono statically by default.

Linking dynamically can make the build a little bit faster (from past experience in Xamarin, we never ported this to .NET when we migrated due to time constraints).

janvorli replied:
The single file is a host statically linked to all the native libraries including coreclr. The scenario when the developers would need static coreclr library would be when they want to use their own host. Thinking about it more, I guess distributing the static coreclr version actually makes sense to enable such scenarios.

rolfbjarne replied:
Yes, that sounds like what we want.

Note that .dylib won't do (Apple doesn't allow them in iOS apps), each dynamic library has to be made into a .framework (which works).

FWIW Apple has recommended using no more than 6-8 .frameworks because otherwise it affects startup performance.

filipnavara commented on src/coreclr/pal/src/exception/seh-unwind.cpp:
```diff
@@ -44,7 +44,7 @@ Module Name:
 #endif // HOST_UNIX
 
-#if defined(TARGET_OSX) && defined(HOST_ARM64) && !defined(HAVE_UNW_AARCH64_X19)
+#if defined(TARGET_APPLE) && defined(HOST_ARM64) && !defined(HAVE_UNW_AARCH64_X19)
```
Fair enough, there's at least 147 usages of `__APPLE__` in CoreCLR vs. 71 of `TARGET_APPLE` (in this PR). I can unify it on `__APPLE__` if that's preferred.

janvorli replied:
Yes, keeping uniformity is preferred. In some later PR, we can later change everything to `TARGET_APPLE`, but doing it now would unnecessarily grow the already large number of files changed in this PR and do that in places that are unrelated to the gist of this PR.

filipnavara replied:
Done

filipnavara committed:
Use __APPLE__ in PAL instead of TARGET_APPLE for consistency

filipnavara committed:
Revert changes to native libs and exclude singlefilehost

filipnavara committed:
Revert more singlefilehost changes

filipnavara committed:
Remove excessive -D_XOPEN_SOURCE

janvorli commented on src/coreclr/pal/src/thread/context.cpp:
```diff
@@ -2181,6 +2184,8 @@ DBG_FlushInstructionCache(
 
     syscall(__NR_riscv_flush_icache, (char *)lpBaseAddress, (char *)((INT_PTR)lpBaseAddress + dwSize), 0 /* all harts */);
+#elif defined(HOST_APPLE) && !defined(HOST_OSX)
+    sys_icache_invalidate((void *)lpBaseAddress, dwSize);
```
Can we use this on all Apple platforms?

filipnavara replied:
Probably yes. I'll check.

filipnavara replied:
It seems that Clang optimizes/inlines `sys_icache_invalidate` on macOS so there's no additional overhead to using it. The same pattern is already present in NativeAOT PAL. I can submit a separate small PR to fix/unify both of these places.

janvorli replied:
I have just tried to compile a simple testing C app with `sys_icache_invalidate` and `__builtin___clear_cache` on my M1 mac. It actually didn't inline anything, and the `__builtin___clear_cache` code just tail called `sys_icache_invalidate`. 
However, on mac x64, the `__builtin___clear_cache` compiles to nothing while the `sys_icache_invalidate` is still a call.

filipnavara replied:
> However, on mac x64, the `__builtin___clear_cache` compiles to nothing while the `sys_icache_invalidate` is still a call.

That's what I originally expected and why I left the `__builtin___clear_cache` in place.

I misinterpreted the output of my test case because `objdump -S` didn't print the correct method names in the object file.

janvorli replied:
Let's leave this change as is then.

filipnavara commented on src/coreclr/CMakeLists.txt:
```diff
@@ -29,7 +29,7 @@ if(CORECLR_SET_RPATH)
 endif(CORECLR_SET_RPATH)
 
-if(CLR_CMAKE_HOST_MACCATALYST OR CLR_CMAKE_HOST_IOS OR CLR_CMAKE_HOST_TVOS)
+if(CLR_CMAKE_HOST_TVOS)
```
I had the same thought. However, .dylibs may still be useful for some simulator scenarios, and there was some build issue when I excluded it. I'll have to run the build again to get more details.

filipnavara replied:
The build fails here when standalone GC is disabled:
```
  ld: library 'gc_pal' not found
  clang++: error: linker command failed with exit code 1 (use -v to see invocation)
  make[2]: *** [dlls/mscoree/coreclr/libcoreclr.dylib] Error 1
```

janvorli replied:
That failure is a bug in the build scripts then. I can see what's causing that. Let's keep this change as is then, I'll look into fixing the build without the standalone GC in a separate PR.

janvorli replied:
I have created a PR with a fix: #111682

filipnavara replied:
Rebased and reverted back to FEATURE_STANDALONE_GC=0.

filipnavara commented on src/coreclr/CMakeLists.txt:
```diff
@@ -59,7 +59,7 @@ include(components.cmake)
 # Build the single file host
 #---------------------------
-if(NOT CLR_CROSS_COMPONENTS_BUILD)
+if(NOT CLR_CROSS_COMPONENTS_BUILD AND NOT CLR_CMAKE_TARGET_MACCATALYST AND NOT CLR_CMAKE_TARGET_IOS AND NOT CLR_CMAKE_TARGET_TVOS)
```
The iOS/macOS workload uses a custom host and not this one. It's definitely possible to build it but there's no use case for it.

janvorli replied:
Ok

filipnavara committed:
Make it possible to compile CoreCLR on iOS-like platforms with './build.sh clr+clr.nativeaotlibs+libs+packs -os [iossimulator/maccatalyst/ios] -arch arm64 -cross'

Build the singlefilehost even if we don't currently use it

Make coreclrhost.h compatible v C(not++)

Fix TargetOS::IsMacOS

WIP: AppleAppBuilder support for CoreCLR

Make the sample app working: dotnet publish src/mono/sample/iOS/Program.csproj -c Release /p:TargetOS=iossimulator /p:TargetArchitecture=arm64 /p:AppleAppBuilderRuntime=CoreCLR /p:DeployAndRun=true /p:UseMonoRuntime=false /p:RunAOTCompilation=false /p:MonoForceInterpreter=false

Cleanup

Update src/tasks/AppleAppBuilder/AppleAppBuilder.csproj

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

Use sys_icache_invalidate for all iOS-like platforms

Consolidate macOS-only entrypoints in a single #if block

Apply feedback

Fix jitinterface cross-build

Infer AppleAppBuilderRuntime from UseMonoRuntime and UseNativeAOTRuntime

Minor changes to test infrastructure

Add basic documentation

Add nearly complete support for building runtime tests

Fix building seh-unwind with older iOS SDK

Use enum for TargetRuntime in AppleAppBuilder

Fix LibraryBuilder build

Cleanup

Fix after rebase

Revert accidental change

Fix tvOS build

Drop CoreCLR tvOS Simulator support for now

filipnavara committed:
Fix build with FEATURE_PERFTRACING_STANDALONE_PAL (debug libs)

janvorli reviewed (approved):
LGTM, thank you!

janvorli commented:
@filipnavara can you please resolve the conflicts so that we can merge the change in?

filipnavara committed:
Build with FEATURE_STANDALONE_GC=0

filipnavara commented:
> @filipnavara can you please resolve the conflicts so that we can merge the change in?

Done.

janvorli commented:
@filipnavara there are some test failures, I think that I've seen some of them in the previous runs already. Could you please take a look to see if they could possibly be related to this change?

filipnavara commented:
- The tvOS failures seems to be https://github.com/dotnet/runtime/issues/108832. The issue is closed but without any link to a resolution. The failure is happening with MonoVM and not in any code that this PR touched.
- wasi-wasm is infrastructure issue.
- osx-arm64 failed with `The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/Microsoft.NETCore.App.MonoCrossAOT/Release/net10.0/osx-arm64/Microsoft.NETCore.App.MonoCrossAOT.deps.json' because it is being used by another process.`. I think it would succeed on a re-run. I don't see an open issue for this.
- windows-x64 failed with `Could not read lines from file "D:\a\_work\1\s\artifacts\obj\Microsoft.NETCore.App.MonoCrossAOT\Release\net10.0\win-x64\Microsoft.NETCore.App.MonoCrossAOT.sfxproj.FileListAbsolute.txt". The process cannot access the file 'D:\a\_work\1\s\artifacts\obj\Microsoft.NETCore.App.MonoCrossAOT\Release\net10.0\win-x64\Microsoft.NETCore.App.MonoCrossAOT.sfxproj.FileListAbsolute.txt' because it is being used by another process.`. Seems similar enough to the above?

I don't think any of that is related to this PR but it may be worth re-running the affected pipelines.

filipnavara commented:
Issue was filed for the MonoCrossAOT build errors - https://github.com/dotnet/runtime/issues/111474. It happened in an unrelated PR.

jkotas commented:
@filipnavara Could you please resolve the conflict?

filipnavara commented:
> @filipnavara Could you please resolve the conflict?

Done

filipnavara commented:
Rebased on top of dotnet/main. Dropped https://github.com/filipnavara/runtime/commit/1094de761de079602cc29726d04a1041a1ad7d1d which I accidentally included in one of the merges and which is necessary until we have a proper build fix for `FEATURE_STANDALONE_GC=0`.

ivanpovazan commented on eng/Subsets.props:
```diff
@@ -92,7 +92,7 @@
     <DefaultCoreClrSubsets Condition="'$(PackageRID)' == 'linux-armel'">clr.native+linuxdac+clr.corelib+clr.tools+clr.nativecorelib+clr.packages+clr.nativeaotlibs+clr.crossarchtools</DefaultCoreClrSubsets>
     <!-- Even on platforms that do not support the CoreCLR runtime, we still want to build ilasm/ildasm. -->
-    <DefaultCoreClrSubsets Condition="'$(PrimaryRuntimeFlavor)' != 'CoreCLR'">clr.iltools+clr.packages</DefaultCoreClrSubsets>
+    <DefaultCoreClrSubsets Condition="'$(RuntimeFlavor)' != 'CoreCLR'">clr.iltools+clr.packages</DefaultCoreClrSubsets>
```
/cc: @steveisok

ivanpovazan commented on src/tests/Directory.Build.props:
```diff
@@ -223,7 +223,8 @@
     <_targetOS>$(TargetOS)</_targetOS>
     <_targetArchitecture>$(TargetArchitecture)</_targetArchitecture>
-    <UseNativeAOTRuntime>true</UseNativeAOTRuntime>
+    <AppleAppBuilderRuntime Condition="'$(TestBuildMode)' == 'nativeaot'">NativeAOT</AppleAppBuilderRuntime>
+    <AppleAppBuilderRuntime Condition="'$(TestBuildMode)' != 'nativeaot'">CoreCLR</AppleAppBuilderRuntime>
```
While it is fine to use: `Runtime` + `targetRuntime` variables instead of `UseNativeAOTRuntime` inside of the AppleAppBuilder, I am a little bit worried about changes around `UseNativeAOTRuntime` as an MSBuild property. /cc: @kotlarmilos

ivanpovazan commented:
@filipnavara regarding my comments on AppBuilder changes, since I already did some work in the similar fashion for Android, let me know if you would like me to resolve them.

filipnavara commented:
> let me know if you would like me to resolve them.

I'm fine either way. You may have better idea how to address this and so it would perhaps be quicker.

My only concern is whether to do it as part of this PR or a follow-up. There are some changes in this PR that are constant source of conflicts (TARGET_OSX -> TARGET_APPLE replacement), and there's a small piece of code that should be landed to unblock eventpipe build for CoreCLR/Android.

filipnavara commented on src/tasks/AppleAppBuilder/TargetRuntime.cs:
```diff
@@ -0,0 +1,9 @@
+
+internal enum TargetRuntime
+{
+    MonoVM,
```
I don't have strong opinion on this and it's fine to change. (It's not the same set of values as RuntimeFlavor though; I think I mirrored it from some other pre-existing place)

filipnavara commented:
Some of the failures in the extra pipelines seem relevant. I'll try to reproduce the `System.Exception: Library not found at: ...` error tomorrow.

filipnavara commented on src/tests/build.proj:
```diff
@@ -454,6 +456,10 @@
       <ExtraAppLinkerArgs Include="@(LinkerArg)" />
     </ItemGroup>
 
+    <PropertyGroup>
+      <NativeDependencies>$(IntermediateOutputPath)\..\$(TestRelativePath)\$(TestName)\native\$(TestName).o</NativeDependencies>
+    </PropertyGroup>
+
```
This seems like duplicate of the same code 10 lines above.

filipnavara committed:
Remove extra NativeDependencies property (bad merge)

ivanpovazan commented on src/tests/build.proj:
```diff
@@ -456,10 +456,6 @@
     </ItemGroup>
 
-    <PropertyGroup>
-      <NativeDependencies>$(IntermediateOutputPath)\..\$(TestRelativePath)\$(TestName)\native\$(TestName).o</NativeDependencies>
```
Is this a fix for failing extra platforms pipeline?

filipnavara replied:
Hopefully yes. I didn't test it locally yet.

ivanpovazan commented:
/azp run runtime-extra-platforms

filipnavara committed:
Add back the UseNativeAOTRuntime property to unblock CI

kotlarmilos commented:
/azp run runtime-extra-platforms

ivanpovazan commented on src/mono/msbuild/apple/build/AppleBuild.targets:
```diff
@@ -298,8 +298,15 @@
       <LinkerArg Remove="@(_LinkerFlagsToDrop)" />
       <ExtraAppLinkerArgs Include="@(LinkerArg)" />
     </ItemGroup>
+ 
+    <PropertyGroup>
+      <AppleAppBuilderRuntime Condition="'$(AppleAppBuilderRuntime)' == '' and '$(UseNativeAOTRuntime)' == 'true'">NativeAOT</AppleAppBuilderRuntime>
+      <AppleAppBuilderRuntime Condition="'$(AppleAppBuilderRuntime)' == '' and '$(UseMonoRuntime)' == 'false'">CoreCLR</AppleAppBuilderRuntime>
+      <AppleAppBuilderRuntime Condition="'$(AppleAppBuilderRuntime)' == ''">MonoVM</AppleAppBuilderRuntime>
+    </PropertyGroup>
```
> I think there was some order dependency

Might be related to when/if `UseNativeAOTRuntime` and `UseMonoRuntime` are set, before importing `AppleBuild.props`. I still think we should move setting these properties there. If the change would break something then we need to fix it - import `AppleBuild.props` early enough in the build process

filipnavara commented:
CI has few timeouts and couple of known issues but the rest passed.

ivanpovazan reviewed (approved):
LGTM! Thanks a lot for all the hard work!

I resolved all my comments related to build integration files and sample changes as I will cover them in a separate PR as part of: https://github.com/dotnet/runtime/issues/111745

steveisok reviewed (approved):
Thanks!

kotlarmilos committed:
Merge branch 'main' into coreclr-ioslike-2

kotlarmilos reviewed (approved):
LGTM!

kotlarmilos commented:
/azp run runtime-extra-platforms

kotlarmilos commented:
The official builds have passed.

akoeplinger commented on src/tasks/AppleAppBuilder/Templates/runtime-coreclr.m:
```diff
@@ -0,0 +1,161 @@
+
+#import "util.h"
+
+#define APPLE_RUNTIME_IDENTIFIER "iossimulator-arm64"
```
this should use the template string like runtime.m

filipnavara replied:
Good catch. I assume we can handle that as part of the follow-ups tracked in https://github.com/dotnet/runtime/issues/111745.

akoeplinger commented on src/coreclr/pal/src/thread/process.cpp:
```diff
@@ -2212,6 +2210,9 @@ PROCCreateCrashDump(
     INT cbErrorMessageBuffer,
     bool serialize)
 {
+#if defined(TARGET_IOS)
```
we should disable this on tvos too

filipnavara replied:
I didn't enable tvOS in this PR. It requires more work and testing due to lack of Mach signal API.

(Previous versions of this PR enabled the tvOS compilation at one point. I may resubmit them separately at some point for a more thorough review. Mixing them with the rest of the iOS bring up made it unreviewable.)

akoeplinger replied:
yes I just wanted to call out so we don't forget to update this spot

akoeplinger reviewed (approved):
LGTM apart from two small comments that can be addressed in a follow-up PR, thanks!

vitek-karas commented:
Thanks a lot Filip - this was a lot of work!!!

mikem8361 commented on src/coreclr/debug/daccess/CMakeLists.txt:
```diff
@@ -47,10 +47,10 @@ target_link_libraries(daccess PRIVATE cdacreader_api)
 add_dependencies(daccess eventing_headers)
 
-if(CLR_CMAKE_HOST_FREEBSD OR CLR_CMAKE_HOST_NETBSD OR CLR_CMAKE_HOST_SUNOS)
+if(CLR_CMAKE_HOST_FREEBSD OR CLR_CMAKE_HOST_NETBSD OR CLR_CMAKE_HOST_SUNOS OR CLR_CMAKE_HOST_APPLE)
```
@hoyosjs, I think it was this change that broke debugging/diagnostics repo tests on MacOS x64. Adding CLR_CMAKE_HOST_APPLE here will cause the runtime DAC table to be generated in the old "rva" way.

filipnavara replied:
I don't think the change was intentional. I traced it back to the first commit on my previous branch but it could have been a result of some rebase.

mikem8361 replied:
Do you want to remove/fix this?

filipnavara replied:
Seems like at some point after the initial patch the conditional compilation of machoreader.cpp was added. This still uses the old `CLR_CMAKE_TARGET_OSX` condition. I'll send a PR to fix this.

filipnavara replied:
https://github.com/dotnet/runtime/pull/112076

ivanpovazan replied:
> @hoyosjs, I think it was this change that broke debugging/diagnostics repo tests on MacOS x64. Adding CLR_CMAKE_HOST_APPLE here will cause the runtime DAC table to be generated in the old "rva" way.

Thanks Mike for drawing attention to this! 

Which pipeline is running these tests? 
I am thinking about ways to improve test coverage for PRs like this, so that we can catch such regressions earlier.

mikem8361 replied:
These failing tests are part of the diagnostics repo.  There is a manual way of running them against a local runtime build, but there currently no way for the runtime repo to use them.  There is Jeremy created this issue https://github.com/dotnet/diagnostics/issues/5213 to track this work.

ivanpovazan replied:
Thanks a lot for the pointers.

mikem8361 commented on src/coreclr/dlls/mscordac/CMakeLists.txt:
```diff
@@ -31,9 +31,9 @@ else(CLR_CMAKE_HOST_WIN32)
-    endif(CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_FREEBSD OR CLR_CMAKE_HOST_NETBSD OR CLR_CMAKE_HOST_SUNOS OR CLR_CMAKE_HOST_HAIKU)
+    endif(CLR_CMAKE_HOST_APPLE OR CLR_CMAKE_HOST_FREEBSD OR CLR_CMAKE_HOST_NETBSD OR CLR_CMAKE_HOST_SUNOS OR CLR_CMAKE_HOST_HAIKU)
 
     if(CORECLR_SET_RPATH AND CLR_CMAKE_HOST_OSX AND CLR_CMAKE_HOST_ARCH_ARM64)
```
Has CLR_CMAKE_HOST_OSX been completed replaced by CLR_CMAKE_HOST_APPLE? If so, then the RPATH in the DAC isn't be added.


@hoyosjs

mikem8361 replied:
Never mind.  CLR_CMAKE_HOST_OSX is still defined if not maccatalyst.

filipnavara replied:
The idea is to use same parameters for all Apple platforms (macOS, iOS, tvOS) whenever possible. RPATH may still need some tweaks for each platform due to different bundle structure on macOS/MacCatalyst and iOS/tvOS.

filipnavara replied:
I'll need to do another pass over the `CLR_CMAKE_HOST_OSX` usages. The RPath ones are likely wrong for some platforms. We don't produce packages for those platforms yet, so the code has never been tested. As for other occurrences - `createdump` is macOS only (but maybe MacCatalyst too?), guards in test code are likely unnecessary because `_DARWIN_C_SOURCE` is now set globally (to be verified).
